[SPARK-5253] [ML] LinearRegression with L1/L2 (ElasticNet) using OWLQN

dbtsai opened the pull request.

dbtsai committed:
first commit

SparkQA commented:
[Test build #26283 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26283/consoleFull) for   PR 4259 at commit [`179f4b3`](https://github.com/apache/spark/commit/179f4b306014273ca7e39878a04ca7ff4d13776f).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26284 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26284/consoleFull) for   PR 4259 at commit [`db8efff`](https://github.com/apache/spark/commit/db8efff6094b93b2c48f10d196f2eb5ca0119feb).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26284 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26284/consoleFull) for   PR 4259 at commit [`db8efff`](https://github.com/apache/spark/commit/db8efff6094b93b2c48f10d196f2eb5ca0119feb).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegression extends Estimator[LinearRegressionModel] with LinearRegressionParams `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26284/
Test FAILed.

SparkQA commented:
[Test build #26283 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26283/consoleFull) for   PR 4259 at commit [`179f4b3`](https://github.com/apache/spark/commit/179f4b306014273ca7e39878a04ca7ff4d13776f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26283/
Test FAILed.

SparkQA commented:
[Test build #26291 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26291/consoleFull) for   PR 4259 at commit [`5b75cda`](https://github.com/apache/spark/commit/5b75cdade0c91db8ca30b2f4012bdeecf8451310).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26292/consoleFull) for   PR 4259 at commit [`aaaaa29`](https://github.com/apache/spark/commit/aaaaa297602fe5278aee1ffcff01ae9cc5c70071).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26291 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26291/consoleFull) for   PR 4259 at commit [`5b75cda`](https://github.com/apache/spark/commit/5b75cdade0c91db8ca30b2f4012bdeecf8451310).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26291/
Test FAILed.

SparkQA commented:
[Test build #26292 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26292/consoleFull) for   PR 4259 at commit [`aaaaa29`](https://github.com/apache/spark/commit/aaaaa297602fe5278aee1ffcff01ae9cc5c70071).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26292/
Test FAILed.

SparkQA commented:
[Test build #26296 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26296/consoleFull) for   PR 4259 at commit [`dc7dcd7`](https://github.com/apache/spark/commit/dc7dcd7d7cbf862a0842fb662b8d2b43d0457d7d).
 * This patch merges cleanly.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26296/
Test FAILed.

shaneknapp commented:
i will kick this off again once i restart jenkins.

shaneknapp commented:
jenkins, test this please.

SparkQA commented:
[Test build #26298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26298/consoleFull) for   PR 4259 at commit [`dc7dcd7`](https://github.com/apache/spark/commit/dc7dcd7d7cbf862a0842fb662b8d2b43d0457d7d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26298/consoleFull) for   PR 4259 at commit [`dc7dcd7`](https://github.com/apache/spark/commit/dc7dcd7d7cbf862a0842fb662b8d2b43d0457d7d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26298/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/param/sharedParams.scala:
```diff
@@ -23,6 +23,27 @@ private[ml] trait HasRegParam extends Params {
 
+private[ml] trait HasElasticNetParam extends HasRegParam {
+  /** param for elastic net regularization parameter */
+  val alphaParam: DoubleParam = new DoubleParam(
```
I would move it to `LinearRegression` and rename it to `alpha`. This is not a shared param.

dbtsai replied:
LOR will have it eventually as well. Do you mean we should `with HasAlpha with HasRegParam` instead of `with HasElasticNetParam`?

jkbradley replied:
I'd vote for not using HasRegParam at all and using elastic net-specific terminology such as l1RegParam, l2RegParam or something else which makes it clear which param applies to which part of the regularization.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/param/sharedParams.scala:
```diff
@@ -23,6 +23,27 @@ private[ml] trait HasRegParam extends Params {
+  def getAlphaParam: Double = get(alphaParam)
+}
+
+private[ml] trait HasConvergenceTol extends Params {
```
In `spark.ml`, we try to use short names for parameters. This could be called `tol`, as used in sklearn and matlab.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/param/sharedParams.scala:
```diff
@@ -23,6 +23,27 @@ private[ml] trait HasRegParam extends Params {
+  def getConvergenceTol: Double = get(convergenceTol)
+}
+
+private[ml] trait HasIntercept extends Params {
```
If we don't allow turning off intercept for now, shall we remove it in this PR?

dbtsai replied:
I would like to be able to disable intercept in this PR. Should not be too hard.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -0,0 +1,316 @@
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql._
+import org.apache.spark.sql.api.scala.dsl._
+import org.apache.spark.sql.catalyst.dsl._
```
`catalyst` is not longer needed.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -0,0 +1,169 @@
+     *                           s0
+     * (Intercept)         6.300528
+     * as.numeric.data.V2. 4.701024
+     * as.numeric.data.V3. 7.198257
```
This is great!

mengxr commented:
@dbtsai Thanks for contributing elastic-net! It may be hard to merge this into 1.3 now. There are two pending 1.3 features that are relevant: DataFrame API and Prediction APIs, both of which will conflict with this PR. We may want to wait for them merged first and update this PR to avoid resolve conflicts frequently.

SparkQA commented:
[Test build #26812 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26812/consoleFull) for   PR 4259 at commit [`9066e4c`](https://github.com/apache/spark/commit/9066e4c7cf24d3d41f96d088ecc75dcaae2c13ef).
 * This patch merges cleanly.

dbtsai commented:
I just rebased and addressed the api changes. 

I was thinking about that when we do standardization on the feature, we're actually rescaling on the data which is basically rescaling on the covariant position of the equation. However, we can achieve the same scaling on rescaling on the graidentSum which is contravariant position. Thus, we don't need to apply the scaler on the data which will be much cheaper. This also works for the logistic regression as well. For intercept in LOR, we can deal with it in gradient function instead of using applyBias, so combining both of those two techniques, we don't have to create new dataset.

SparkQA commented:
[Test build #26812 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26812/consoleFull) for   PR 4259 at commit [`9066e4c`](https://github.com/apache/spark/commit/9066e4c7cf24d3d41f96d088ecc75dcaae2c13ef).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegression extends Estimator[LinearRegressionModel] with LinearRegressionParams `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26812/
Test PASSed.

jkbradley commented:
@dbtsai I'd like to make a pass over this, but I realized that it has conflicts because of the developer api PR committed last week: [https://github.com/apache/spark/pull/3637]  Could you please rebase?  I don't think there are any more big PRs coming up which will make you rebase again.  Thank you!

debasish83 commented:
@dbtsai we are close to merge this PR which brings OWLQN and PQN under the umbrella of proximal algorithms to support most of the interesting ML related constraints https://github.com/scalanlp/breeze/pull/364...

I also have another Breeze PR that merges OWLQN and PQN through Proximal Quasi Newton method but that will take time to stabilize..

I would like to add more tests for OWLQN and specifically comparisons with liblinear...Most likely if I find some interesting testcases in your PR, I will move them to NonlinearMinimizerTest...

I will open up a mllib PR that uses NonlinearMinimizer and tests a family of consensus algorithms (Block Coordinate Descent and ADMM)..Our main focus is ElasticNet as well so it will be great if this PR is merged so that I can build upon it...

debasish83 commented:
this is linear regression...what happened to the logistic regression elastic net? We are more interested in that one...

dbtsai commented:
@jkbradley I will rebase soon. @debasish83 I'll add MLOR with elastic-net when we stabilize the new ML api. Doing this in old codebase will be huge effort, and I will like to make them self-contain as much as possible instead of sharing the same generalized linear algorithm base class. We already have enough if else statement to deal with the differences in the base class.

SparkQA commented:
[Test build #29042 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29042/consoleFull) for   PR 4259 at commit [`66a4dc3`](https://github.com/apache/spark/commit/66a4dc31aa56ced603cd1172719dc4510fcdbaa1).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29042 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29042/consoleFull) for   PR 4259 at commit [`66a4dc3`](https://github.com/apache/spark/commit/66a4dc31aa56ced603cd1172719dc4510fcdbaa1).
 * This patch **fails Scala style tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29042/
Test FAILed.

dbtsai commented:
@jkbradley and @mengxr I just rebased it. Will do couple optimizations to avoid the scaling on the datasets which can be done in the optimization instead. You guys can start to give me feedback so we have ample time to address issues before 1.4. Thanks.

SparkQA commented:
[Test build #29043 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29043/consoleFull) for   PR 4259 at commit [`ea3e1dc`](https://github.com/apache/spark/commit/ea3e1dc55583d1fdd69c74a0201c5743a0baef2a).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29044 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29044/consoleFull) for   PR 4259 at commit [`ec3572f`](https://github.com/apache/spark/commit/ec3572fa084d166d541f3d1edd0fc57c52cdade2).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29043 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29043/consoleFull) for   PR 4259 at commit [`ea3e1dc`](https://github.com/apache/spark/commit/ea3e1dc55583d1fdd69c74a0201c5743a0baef2a).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29043/
Test PASSed.

SparkQA commented:
[Test build #29044 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29044/consoleFull) for   PR 4259 at commit [`ec3572f`](https://github.com/apache/spark/commit/ec3572fa084d166d541f3d1edd0fc57c52cdade2).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29044/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/param/sharedParams.scala:
```diff
@@ -34,6 +34,43 @@ private[ml] trait HasRegParam extends Params {
   def getRegParam: Double = get(regParam)
 }
 
+private[ml] trait HasElasticNetParam extends HasRegParam {
```
Do we want to call this "alphaParam?"  That name assumes people have read elastic net papers.  What about something like:
* "elasticNetParam" (very explicit) (I assume this was your original name.)
* "regMixing" (also explicit)
* "regAlpha" (at least will be grouped next to "regParam")

dbtsai replied:
I voted for `regAlphaParam: Double` for the variance name, and keep `HasElasticNetParam` as the trait name. What do you think?

jkbradley replied:
I think the names should definitely match, so we should pick one.  I'm OK with "elasticNetParam" or "regAlpha."  (I think "regAlpha" doesn't need "Param" attached to it since "alpha" is the name of the parameter.)  Since "elastic net" is more easily recognized than "alpha," I vote for "elasticNetParam."

dbtsai replied:
Sounds great! I'm convinced to do something like

```scala
private[ml] trait HasElasticNetParam extends HasRegParam {
  /**
   * param for elastic net regularization parameter
   * @group param
   */
  val elasticNetParam: DoubleParam =
    new DoubleParam(this, "elasticNetParam", "the ElasticNet mixing parameter")

  /** @group getParam */
  def getElasticNetParam: Double = get(elasticNetParam)
}
```

jkbradley replied:
Nice, I like that it extends HasRegParam.  When you update it, can you please make the doc specify more about the parameter (range, plus which end of the range corresponds to L1 vs. L2)?

dbtsai replied:
Definitely. It seems that there is no easy way to specify the valid parameter range or requirement in this framework. Do you think it's a good idea that I add the check like the following?

```scala
  def getElasticNetParam: Double = {
    val elasticNetParam = get(elasticNetParam)
    require(elasticNetParam >= 0)
    require(elasticNetParam <= 1)
    elasticNetParam
  }
```

jkbradley commented:
@dbtsai Thanks for the update!  I'll make a more detailed pass soon.

One bigger question: What is the plan for extensions?  Based on this PR, it sounds like we expect "LinearRegression" to only support no regularization, L1, L2, or elastic net.  I assume that, in the future, users who want other regularization can write their own extensions of GeneralizedLinearAlgorithm (once that is in spark.ml).  But would we ever want to add more built-in regularizers?  Or are we deciding now that we never will?

dbtsai commented:
@jkbradley I think we should only support basic regularization in spark.ml first which is what python scikit-learn does. If users have the need of different type of regularization, they can implement it based on the code we have. 

It will be hard to implement GeneralizedLinearAlgorithm with regularization without using a lot of if-else statement to handle the special case. I implemented logistic regression, linear regression, and cox proportional-hazards regression with elasticnet regularization at Alpine, and our customers are asking for precise accuracy compared with R's glmnet package. As a result, I spent some time to research the original R's glmnet code, and I found that there is no generic way to handle different linear models. There are special cases here and there.

For example, in logistic regression, the intercept is computed by adding extra one dimension in the data with constant one, but in linear regression, the intercept is computed by `val intercept = yMean - dot(weights, scaler.mean)`.

As a result, I would like to implement them separately and make sure we have the same accuracy compared with R with proper tests first, and then we can abstract out the common part. I have another PR trying to do this, #1518 and I will continuous on that after this PR is merged.

jkbradley commented:
@dbtsai  That sounds reasonable.  I think we could do a general implementation based on abstractions for regularization and loss functions, but it might not be as efficient as specializations (such as closed-form computation of the intercept for linear regression).  I'm OK with this for now, and I definitely support comparisons with R for correctness.

Let me know when it's all ready so I can make a detailed pass!

SparkQA commented:
[Test build #29327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29327/consoleFull) for   PR 4259 at commit [`14fd603`](https://github.com/apache/spark/commit/14fd603f84a92896ffec877bf24d47d303d094ae).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29327/consoleFull) for   PR 4259 at commit [`14fd603`](https://github.com/apache/spark/commit/14fd603f84a92896ffec877bf24d47d303d094ae).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29327/
Test PASSed.

SparkQA commented:
[Test build #29990 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29990/consoleFull) for   PR 4259 at commit [`ea8d1a4`](https://github.com/apache/spark/commit/ea8d1a44d43483ba1c21beec74b3bb5ea5c86c12).

SparkQA commented:
[Test build #29990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29990/consoleFull) for   PR 4259 at commit [`ea8d1a4`](https://github.com/apache/spark/commit/ea8d1a44d43483ba1c21beec74b3bb5ea5c86c12).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29990/
Test PASSed.

SparkQA commented:
[Test build #30508 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30508/consoleFull) for   PR 4259 at commit [`825ca02`](https://github.com/apache/spark/commit/825ca0274003f841e2e6c67f95d0bcdc5bbdbb2d).

SparkQA commented:
[Test build #30508 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30508/consoleFull) for   PR 4259 at commit [`825ca02`](https://github.com/apache/spark/commit/825ca0274003f841e2e6c67f95d0bcdc5bbdbb2d).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.
 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30508/
Test FAILed.

SparkQA commented:
[Test build #30510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30510/consoleFull) for   PR 4259 at commit [`c867838`](https://github.com/apache/spark/commit/c8678386eba64938c647e8bcca060cd68e884e2c).

SparkQA commented:
[Test build #30510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30510/consoleFull) for   PR 4259 at commit [`c867838`](https://github.com/apache/spark/commit/c8678386eba64938c647e8bcca060cd68e884e2c).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.
 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30510/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30608/
Test FAILed.

dbtsai committed:
Refactoring

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala:
```diff
@@ -256,4 +256,38 @@ trait HasFitIntercept extends Params {
+ * Trait for shared param elasticNetParam.
+ */
+@DeveloperApi
+trait HasElasticNetParam extends Params {
```
This shouldn't be a shared param.

dbtsai replied:
do you suggest to move it to LinearRegression.scala? we will use it in LOR as well.

mengxr replied:
okay

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
 class LinearRegression extends Regressor[Vector, LinearRegression, LinearRegressionModel]
   with LinearRegressionParams {
 
-  setDefault(regParam -> 0.1, maxIter -> 100)
```
minor: I think it is okay to leave the defaults here. It is easier for users to read. We put some below params because we need to have the params defined first.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+   * @group setParam
+   */
   def setRegParam(value: Double): this.type = set(regParam, value)
+  setDefault(regParam -> 0.0)
```
Could we set something like `0.01` with L2? Having zero regularization is not a good practice.

dbtsai replied:
To match R's default result, we need to `0.0`. Also, the meaning of lambda will be changed if the numbers of sample is changed. So it's hard to come out with a good default. Why don't we implement regularization path to find the best lambda?

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+   * @group setParam
+   */
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-9)
```
Do we need `1e-9` by default? I'm thinking about some value like `1e-6`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+    val effectiveL2RegParam = (1.0 - paramMap(elasticNetParam)) * effectiveRegParam
+
+    val costFun = new LeastSquaresCostFun(
+      instances,
```
We don't chop down args in a function call. It should be 

~~~scala
val costFun = new ... (instances, yStd, ....
  effectiveL2RegParam)
~~~

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+
+    val initialWeights = Vectors.zeros(numFeatures)
+    val states =
+      optimizer.iterations(new CachedDiffFunction(costFun), initialWeights.toBreeze.toDenseVector)
```
`initialWeights` is already a dense vector.

dbtsai replied:
`optimizer.iterations` explicitly takes `BDV`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+      optimizer.iterations(new CachedDiffFunction(costFun), initialWeights.toBreeze.toDenseVector)
+
+    var state = states.next()
+    val lossHistory = new ArrayBuffer[Double](paramMap(maxIter))
```
minor. It would be simpler to use `ArrayBuilder` without preallocated space. User may set `maxIter` to a very large number just to wait until convergence.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+
+    var state = states.next()
+    val lossHistory = new ArrayBuffer[Double](paramMap(maxIter))
+    while(states.hasNext) {
```
space after `while`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -42,34 +50,122 @@ private[regression] trait LinearRegressionParams extends RegressorParams
+      val rawWeights = state.x.toArray.clone()
+      var i = 0
+      while (i < rawWeights.length) {
+        rawWeights(i) = if (featuresStd(i) != 0.0) rawWeights(i) * yStd / featuresStd(i) else 0.0
```
This may be more readable:

~~~scala
rawWeights(i) *= if (featuresStd(i) != 0.0) yStd / featuresStd(i) else 0.0
~~~

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -97,3 +193,153 @@ class LinearRegressionModel private[ml] (
   }
 }
+
+private class LeastSquaresAggregator(
```
Add doc.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -97,3 +193,153 @@ class LinearRegressionModel private[ml] (
+    require(dim == data.size, s"Dimensions mismatch when adding new sample." +
+      s" Expecting $dim but got ${data.size}.")
+
+    val diff = dot(data, effectiveWeightsVector) - (label - labelMean) / labelStd + offset
```
`- (label - labelMean) / labelStd + offset = - label / labelStd + (label/Mean / labelStd + offset)`

Shall we cache the second part that is a constant?

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -97,3 +193,153 @@ class LinearRegressionModel private[ml] (
+      diffSum += other.diffSum
+
+      var i = 0
+      val localThisGradientSumArray = this.gradientSumArray
```
This optimization is necessary, since this is not on the critical path. The following should be suffcient:

~~~scala
if (other.totalCnt != 0) {
  totalCnt += other.totalCnt
  lossSum += other.diffSum
  System.arraycopy(other.gradientSumArray, 0, this.gradientSumArray, 0, dim)
}
~~~

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -97,3 +193,153 @@ class LinearRegressionModel private[ml] (
+      Vectors.dense(temp)
+    }
+
+    axpy(-diffSum, result, correction)
```
Is `correction` used?

dbtsai replied:
This is a bug. The unit test doesn't catch it since `diffSum` is too small, so we are basically adding zeros into `result`. Let me try to come out with another synthetic dataset which can catch this.

dbtsai committed:
addressed feedback and fixed a bug. TODO: documentation and build another
synthetic dataset which can catch the bug fixed in this commit.

SparkQA commented:
[Test build #31090 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31090/consoleFull) for   PR 4259 at commit [`b46679c`](https://github.com/apache/spark/commit/b46679c56f3d2d19a6103b5c5cc28e566880b086).

SparkQA commented:
[Test build #31090 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31090/consoleFull) for   PR 4259 at commit [`b46679c`](https://github.com/apache/spark/commit/b46679c56f3d2d19a6103b5c5cc28e566880b086).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasElasticNetParam extends Params `
  * `trait HasTol extends Params `

 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.7.jar`
   * `breeze-macros_2.10-0.11.2.jar`
   * `breeze_2.10-0.11.2.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.4.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.7.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

dbtsai committed:
add comments

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -97,3 +190,146 @@ class LinearRegressionModel private[ml] (
+    }
+
+    axpy(-diffSum, correction, result)
+    scal(1.0 / totalCnt, result)
```
Okay, I finally found why `correction` effect is zero. It's because `diffSum` is zero in our test dataset. `diffSum` is sum of `diff`, and for a synthetic dataset generated from linear equation with noise, the average of `diff` will be zero. As a result, for a real non-linear dataset, `diffSum` will not be zero, so we need some non-linear dataset for testing correctness. I'll add famous prostate cancer dataset used in the linear regression lasso paper into the unit-test.

dbtsai committed:
rebase

SparkQA commented:
[Test build #31130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31130/consoleFull) for   PR 4259 at commit [`9fc48ed`](https://github.com/apache/spark/commit/9fc48edfe420499014b239f87a67116a4a23782a).

SparkQA commented:
[Test build #31130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31130/consoleFull) for   PR 4259 at commit [`9fc48ed`](https://github.com/apache/spark/commit/9fc48edfe420499014b239f87a67116a4a23782a).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasElasticNetParam extends Params `
  * `trait HasTol extends Params `

 * This patch does not change any dependencies.

dbtsai committed:
add import org.apache.spark.util.Utils back

SparkQA commented:
[Test build #31133 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31133/consoleFull) for   PR 4259 at commit [`a81c201`](https://github.com/apache/spark/commit/a81c201d61d6952e5cb0c3682760b136ac8297ba).

SparkQA commented:
[Test build #31133 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31133/consoleFull) for   PR 4259 at commit [`a81c201`](https://github.com/apache/spark/commit/a81c201d61d6952e5cb0c3682760b136ac8297ba).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasElasticNetParam extends Params `
  * `trait HasTol extends Params `

 * This patch does not change any dependencies.

mengxr commented:
LGTM. Merged into master. Thanks!!
